core/src/sumeragi/main_loop.rs

```diff
@@ -378,377 +406,475 @@ impl Sumeragi {
                 let _ = voting_block.take();
-                if let Some(mut new_block) =
-                    self.vote_for_block(state, current_topology, genesis_public_key, block_created)
+
+                if let Some(mut valid_block) =
+                    self.validate_block(state, &self.topology, genesis_account, block_created)
                 {
                     // NOTE: Up until this point it was unknown which block is expected to be received,
                     // therefore all the signatures (of any hash) were collected and will now be pruned
-                    add_signatures::<false>(&mut new_block, voting_signatures.drain(..));
-                    *voting_block = Some(new_block);
+
+                    for signature in core::mem::take(voting_signatures) {
+                        if let Err(error) =
+                            valid_block.block.add_signature(signature, &self.topology)
+                        {
+                            debug!(?error, "Signature not valid");
+                        }
+                    }
+
+                    *voting_block = self.try_commit_block(valid_block, is_genesis_peer);
+                } else {
+                    // FIX: Restore `voting_block`
+                    // Currently it's not possible because block writer needs to be released
+                    // Look at https://github.com/hyperledger/iroha/issues/4643
                 }
             }
-            (BlockMessage::BlockSigned(BlockSigned { hash, signatures }), Role::ProxyTail) => {
-                trace!(block_hash=%hash, "Received block signatures");
+            (BlockMessage::BlockSigned(BlockSigned { signature }), Role::ProxyTail) => {
+                info!(
+                    peer_id=%self.peer_id,
+                    role=%self.role(),
+                    "Received block signatures"
+                );
 
-                let roles: &[Role] = if current_view_change_index >= 1 {
-                    &[Role::ValidatingPeer, Role::ObservingPeer]
+                if let Ok(signatory_idx) = usize::try_from(signature.0) {
+                    let signatory = &self.topology.as_ref()[signatory_idx];
+
+                    match self.topology.role(signatory) {
+                        Role::Leader => error!(
+                            peer_id=%self.peer_id,
+                            role=%self.role(),
+                            "Signatory is leader"
+                        ),
+                        Role::Undefined => error!(
+                            peer_id=%self.peer_id,
+                            role=%self.role(),
+                            "Unknown signatory"
+                        ),
+                        Role::ObservingPeer if view_change_index == 0 => error!(
+                            peer_id=%self.peer_id,
+                            role=%self.role(),
+                            "Signatory is observing peer"
+                        ),
```

Comment:
I'd like to take this opportunity to ask, why after view changes does the proxy tail accept block signatures from _observing_ peers as well as validating peers? Shouldn't the roles change along with the view changes?

Reply:
I think this was done as optimization, to reduce amount of view changes required to push faulty peer from voting set.

Reply:
Ok, I can see #3110

Reply:
> Shouldn't the roles change along with the view changes?

I think they should but the current implementation is such that they don't

Reply:
this should be considered in #4346